File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.cpp
class Solution {
public:
    int furthestBuilding(vector<int>& h, int bricks, int ladder) {
        priority_queue<int>q;
        
        int i=0;
        int sum=0;
        for(i=0;i<h.size()-1;i++)
        {
           
            int a=h[i+1]-h[i];
             // cout<<i<<" "<<sum<<" "<<a<<"    ";
            if(a>0 && sum+a<=bricks){
                sum+=a;
                q.push(a);
            }
            else if(a>0 && sum+a>bricks)
            {
                if((q.empty() || a>q.top()) && ladder>0)
                {
                    // cout<<1<<" ";
                    ladder--;
                }
                else if(ladder>0)
                {
                    // cout<<2<<" ";
                    ladder--;
                    sum-=q.top();
                    q.pop();
                    sum=sum+a;
                    q.push(a);
                }
                else
                    break;
            }
            // cout<<3<<" ";
        }
       
        return i;
    }
};

File: 1642-furthest-building-you-can-reach/1642-furthest-building-you-can-reach.java
class Solution {
    public int furthestBuilding(int[] heights, int bricks, int ladders) {
        PriorityQueue<Integer> q=new PriorityQueue<Integer>(new Comparator<Integer>(){
            public int compare(Integer x,Integer y){
                if(x>y){
                    return 1;
                }else{
                    return -1;
                }
            }
        });
        int greaterSum=0;
        int i=0;
        for(i=1;i<heights.length;i++){
            if(heights[i]>heights[i-1]){
                if(q.size()<ladders){
                    q.add(heights[i]-heights[i-1]);
                }else{
                    int useBricks=heights[i]-heights[i-1];
                    if(!(q.isEmpty()) && q.peek()<heights[i]-heights[i-1]){
                        useBricks=q.poll();
                        q.add(heights[i]-heights[i-1]);
                    }
                    if(useBricks>bricks){
                        return i-1;
                    }else{
                        bricks-=useBricks;
                    }
                }
            }
        }
        return i-1;
    }
}